### PR TITLE
SCMOD-6955: Add job type support

### PR DESCRIPTION
**Pull request by joseph-lansdowne-opentext**
- ticket: https://portal.digitalsafe.net/browse/SCMOD-6955
- build: http://sou-jenkins2.hpeswlab.net/job/JobService/job/JobService~job-service~SCMOD-6955~CI/

----

Potential issues:
 - I used a singleton (`JobTypeDefinitions`).  (I would have liked to use a different mechanism, such as storing an instance in jaxrs `ServletContext`, but I don't think I have access to that without modifying the swagger generated code.  I guess another solution would be to add a `JobsPut#setJobTypeDefinitions` static method.)
 - ~~When parsing a job type definition, the code doesn't have access to the filename the definition was loaded from, so if we can't parse the ID there's no identifying information in the error message.~~
 - Job service doesn't exit if it fails loading definitions on startup.  (Not sure how to do this.)
 - There aren't any integration tests for invalid definitions.  (Couldn't think of a way to do this.)

Design notes:
 - In the end, I used JSLT.  There's an argument for using JavaScript, since it's part of the worker framework and is used by the workflow worker.  JSLT has the benefit of not needing to encode or decode JSON any additional times, and the data manipulations we'd use are generally more concise.
 - I used a single YAML file with embedded components, to be similar to the workflow worker.
 - It would be nice to have a framework for testing job type definition files, but maybe that's outside the scope of this ticket.

### CODE REVIEW

File: docs/pages/en-us/Job-Types.md
Comment (dermot-hardy) [inline docs/pages/en-us/Job-Types.md:1]: I looking on the [Features](https://pages.github.houston.softwaregrp.net/sepg-docs-qa/JobService_CI_job-service_SCMOD-6955/pages/en-us/Features) page on the Documentation QA site and I can't see a way to navigate to this new page.  Should it be linked?  Or be listed in the nav. bar on the right-hard side?

File: docs/pages/en-us/Job-Types.md
Comment (dermot-hardy) [inline docs/pages/en-us/Job-Types.md:74]: Elsewhere we've used draft 7.  Why did you decide to use these older versions?

File: job-service-container/pom.xml
Comment (dermot-hardy) [inline job-service-container/pom.xml]: This appears to be bundling files intended for testing into the actual image.

File: docs/pages/en-us/Job-Types.md
Comment (dermot-hardy) [inline docs/pages/en-us/Job-Types.md]: Is this required or could the base filename be used as the id?

File: job-service-container/src/integration-test/resources/job-type-definitions/wrong-ext.notyaml
Comment (dermot-hardy) [inline job-service-container/src/integration-test/resources/job-type-definitions/wrong-ext.notyaml]: Add the license header to this file

File: job-service-container/src/integration-test/java/com/hpe/caf/services/job/api/JobServiceIT.java
Comment (dermot-hardy) [inline job-service-container/src/integration-test/java/com/hpe/caf/services/job/api/JobServiceIT.java]: The convention is to use `static final` rather than `final static`.  Think "science fiction" rather than "fiction science".  (Also you should be consistent with the line above.)

File: job-service-container/src/integration-test/java/com/hpe/caf/services/job/api/JobServiceIT.java
Comment (dermot-hardy) [inline job-service-container/src/integration-test/java/com/hpe/caf/services/job/api/JobServiceIT.java]: Can this method be `static`?

File: job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml
Comment (dermot-hardy) [inline job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: Looking at the way this is defined I'm wondering if we should move it up a level (i.e. and have `type` and `parameters` directly on `newJob`).  What do you think?

File: job-service/src/main/java/com/hpe/caf/services/job/jobtype/JobTypes.java
Comment (dermot-hardy) [inline job-service/src/main/java/com/hpe/caf/services/job/jobtype/JobTypes.java]: `getjobType()` --> `getJobType()`

File: job-service/src/test/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParserTest.java
Comment (dermot-hardy) [inline job-service/src/test/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParserTest.java]: Either change `Collections.emptyMap()` to `Collections.EMPTY_MAP` or add `@SuppressWarnings("AssertEqualsBetweenInconvertibleTypes")` to the method.

File: job-service/src/test/java/com/hpe/caf/services/job/jobtype/DirectoryLoaderTest.java
Comment (dermot-hardy) [inline job-service/src/test/java/com/hpe/caf/services/job/jobtype/DirectoryLoaderTest.java]: Use `Collections.EMPTY_LIST` or `@SuppressWarnings("AssertEqualsBetweenInconvertibleTypes")`.

File: job-service/src/test/java/com/hpe/caf/services/job/jobtype/JobTypeTest.java
Comment (dermot-hardy) [inline job-service/src/test/java/com/hpe/caf/services/job/jobtype/JobTypeTest.java]: You're putting the assert parameters in the wrong order - for JUnit, unlike TestNG, the order is `(expected, actual)`.

File: job-service/src/test/java/com/hpe/caf/services/job/jobtype/JsltTaskDataBuilderTest.java
Comment (dermot-hardy) [inline job-service/src/test/java/com/hpe/caf/services/job/jobtype/JsltTaskDataBuilderTest.java:64]: Add `@SuppressWarnings("ResultOfObjectAllocationIgnored")` to the method.

File: job-service/src/main/java/com/hpe/caf/services/job/jobtype/DirectoryLoader.java
Comment (dermot-hardy) [inline job-service/src/main/java/com/hpe/caf/services/job/jobtype/DirectoryLoader.java]: Since you're using the NIO stuff you should probably use `Files.newInputStream()` instead.

File: job-service/src/main/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParser.java
Comment (dermot-hardy) [inline job-service/src/main/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParser.java:215]: Should we not have a `getDescription()` or is this only to help with debugging?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: docs/pages/en-us/Job-Types.md
Comment (joseph-lansdowne-opentext) [reply docs/pages/en-us/Job-Types.md:74]: This is what the validator I used claims to support.  I chose this validator because it's used elsewhere (repos `worker-elastic`, `worker-document`) and works directly with jackson's `JsonNode`.

File: job-service-container/pom.xml
Comment (joseph-lansdowne-opentext) [reply job-service-container/pom.xml]: Right, so the build part will become the production image but the rest (eg. env, volumes) is for tests?  So I should do this by creating another image like `keystore` and mount it over as a volume?

File: docs/pages/en-us/Job-Types.md
Comment (joseph-lansdowne-opentext) [reply docs/pages/en-us/Job-Types.md]: It could work that way; I don't prefer either way.  In the code I treat the definition source independently from its contents, so it doesn't necessarily have a name, but name could be added as a separate concept that also comes from the source.

File: job-service-container/pom.xml
Comment (dermot-hardy) [reply job-service-container/pom.xml]: That is one way.  The other way is to build an extra testing image layer on top of this one.  Either is fine.

File: job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml
Comment (joseph-lansdowne-opentext) [reply job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: Works for me.  It would save talking about the confusing `job.job`.

File: docs/pages/en-us/Job-Types.md
Comment (dermot-hardy) [reply docs/pages/en-us/Job-Types.md]: I tend to prefer that the name is used as otherwise the filename and id will get out-of-sync and cause confusion.

File: job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml
Comment (dermot-hardy) [reply job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: Ok

File: job-service/src/main/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParser.java
Comment (joseph-lansdowne-opentext) [reply job-service/src/main/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParser.java:215]: I'm not sure you would ever care what description is - the field's only there because the parser fails when it finds a YAML property and nowhere to put the value.  But there's no harm adding it for completeness.

Comment (dermot-hardy) [reply job-service/src/main/java/com/hpe/caf/services/job/jobtype/DefaultDefinitionParser.java:215]: Ok.  Either is fine.

File: job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml
Comment (joseph-lansdowne-opentext) [reply job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: `type` and not `typeId`?

Comment (dermot-hardy) [reply job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: `type` sounds fine to me.

Comment (joseph-lansdowne-opentext) [reply job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: Just comparing to partitions, where you preferred `partitionId`.

Comment (dermot-hardy) [reply job-service-contract/src/main/resources/com/hpe/caf/services/job/swagger.yaml]: True - I guess I see types as closer to an enumeration but it is very subjective.

**Comment by joseph-lansdowne-opentext** [comment]
That should be everything, still building though.

File: docs/pages/en-us/Job-Types.md
Comment (dermot-hardy) [inline docs/pages/en-us/Job-Types.md]: Make these links

**Review by dermot-hardy — APPROVED** [review]

**Comment by dermot-hardy** [comment]
Do you think that the headers should also be made available to the transformation script?

**Comment by joseph-lansdowne-opentext** [comment]
Do you mean all the request headers?  Do you have a use case?  I can't really think of anything that couldn't be either a header handled by the job service itself (eg. auth token, instruction to disallow jobs not using type), or a parameter in the request body.

**Comment by dermot-hardy** [comment]
I was thinking about whether it might be desirable to be able to read something like an "X-TENANT-ID" header.  I see you've made the partition id available, but that does not necessarily directly align to the tenant id.  Indeed we intentionally didn't name it tenant id to be clear about that.  I don't think it's important for now - we can always add the feature later if we need to.

**Comment by dermot-hardy** [comment]
Let's do the 3.0 release ([SCMOD-6893](https://portal.digitalsafe.net/browse/SCMOD-6893)) before merging this and make it a 3.1 feature.  (I'd like there to be more time to test it.)

**Comment by dermot-hardy** [comment]
Actually I've just tried testing this and I can't get the swagger page to come up.  Is it coming up for you OK?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
I found this in the log which probably explains it:

     java.lang.RuntimeException: Error loading job type definitions
        at com.hpe.caf.services.job.jobtype.JobTypesContextListener.contextInitialized(JobTypesContextListener.java:47)
        at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4682)
        ...
        at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:492)
    Caused by: com.hpe.caf.services.job.jobtype.InvalidJobTypeDefinitionException: batch-ingestion-job: target pipe is not configured
        at com.hpe.caf.services.job.jobtype.DefaultDefinitionParser$Definition.getTargetPipe(DefaultDefinitionParser.java:148)
        at com.hpe.caf.services.job.jobtype.DefaultDefinitionParser.parse(DefaultDefinitionParser.java:66)
        at com.hpe.caf.services.job.jobtype.DirectoryLoader.load(DirectoryLoader.java:53)
        at com.hpe.caf.services.job.jobtype.JobTypes.<init>(JobTypes.java:42)
        at com.hpe.caf.services.job.jobtype.JobTypes.initialise(JobTypes.java:63)
        at com.hpe.caf.services.job.jobtype.JobTypesContextListener.contextInitialized(JobTypesContextListener.java:44)
        ... 41 more

But the reason I didn't set a target pipe is because I didn't want one - target pipe should be optional.

**Comment by joseph-lansdowne-opentext** [comment]
Just target pipe?  (Task pipe is also currently required.)

In a way, I think it's nicer to have to set the environment variable for it to startup, because otherwise if you make a typo in the environment variable name it will take longer to discover and track down the issue.  The only sensible value for that is empty string though, and maybe it's weird to require you to set an environment variable to empty.

**Comment by dermot-hardy** [comment]
Yes task pipe is mandatory as the message has to go somewhere.

I'm fine with it having to be present and empty so long as that translates to it not being specified.

**Comment by joseph-lansdowne-opentext** [comment]
Well, the target pipe change is ready, if Jenkins ever finishes building it.

I've also made the errors change I mentioned on the pull request to the other repo.  Unfortunately I couldn't find any better way of distinguishing the error types than this.

**Review by dermot-hardy — APPROVED** [review]

**Comment by dermot-hardy** [comment]
Did the swagger page come up OK for you?

**Comment by joseph-lansdowne-opentext** [comment]
Looks like the `default: null` in the swagger definition makes the UI unhappy, although it works just fine with the client/server generator and the online validator.  It's just a label, so I'll remove it.  (I wonder how swagger UI issues could be picked up by automated testing?)

If this is waiting for the release, what state should I leave the ticket/PR in?  Just as they are?

**Comment by dermot-hardy** [comment]
I don't think there's anything holding up the release.  Check with @michael-bryson.  If there isn't then we can release today and then merge this immediately after release.

**Comment by michael-bryson** [comment]
@dermot-hardy There have been a couple of commits to develop today which have not been QA'd yet and there is also a test failure on the release build so can't be released until that is resolved.

**Comment by dermot-hardy** [comment]
OK thanks.

**Comment by joseph-lansdowne-opentext** [comment]
Merged develop to resolve release notes conflict.

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**
